pkg/objcache/objcache.go

```diff
@@ -112,9 +100,32 @@ func (c *Cache) Create(key string, size int64) (writer io.Writer, err error) {
 			return nil, ErrCacheFull
 		}
 	}
-	c.entries[key] = NewBuffer(make([]byte, 0, int(size)))
-	c.currentSize += valueLen
-	return c.entries[key], nil
+
+	// Will hold the object contents.
+	buf := bytes.NewBuffer(make([]byte, 0, size))
+	// Account for the memory allocated above.
+	c.currentSize += uint64(size)
+	// Implements io.WriteCloser
+	return struct {
+		io.Writer
+		io.Closer
+	}{
+		// Implements io.Writer
+		buf,
+		// Implements io.Closer
+		addToCache(func() error {
+			c.mutex.Lock()
+			defer c.mutex.Unlock()
+			if buf.Len() != int(size) {
+				// Full object not available hence do not save buf to object cache.
+				c.currentSize -= uint64(size)
+				return nil
+			}
+			// Full object available in buf, save it to cache.
+			c.entries[key] = buf.Bytes()
```

Comment:
Did you test the overall memory usage of using this style v/s having our own buffer implementation?